d3dcore/include/d3dcore/d3d_info.h

```diff
@@ -0,0 +1,253 @@
+}
+
+template<typename Container>
+D3DInfo<Container>::~D3DInfo()
+{
+	SAFE_RELEASE(m_dxdiag_provider);
+	SAFE_RELEASE(m_dxdiag_root);
+}
+
+template<typename Container>
+void D3DInfo<Container>::query_info()
+{
+
+	traverse(m_dxdiag_root, m_root_container);
+
+
+}
+
+template<typename Container>
+std::shared_ptr<Container> D3DInfo<Container>::get_root_container()
+{
+	return this->m_root_container;
+}
+
+//}
+
+
+
+
+template<typename Container>
+std::string D3DInfo<Container>::get_property_value(IDxDiagContainer * dxContainer, WCHAR * propName)
+{
+	HRESULT hr;
+	VARIANT var;
+	VariantInit(&var);
+	std::string result;
+
+	if (FAILED(hr = dxContainer->GetProp(propName, &var))) return std::string();
+
+
+	switch (var.vt) {
+
+	case VT_I4:
+		result = std::to_string(var.intVal);
+		break;
+
+	case VT_UI4:
+		result = std::to_string(var.uintVal);
+		break;
+
+	case VT_BOOL:
+		result = var.boolVal != 0 ? std::string("true") : std::string("false");
+		break;
+
+		/*case VT_BSTR:
+
+	#ifdef _UNICODE
+			wcsncpy(strValue, var.bstrVal, 256); //todo: update the length of string
+	#else
+			wcstombs(strValue, var.bstrVal, nStrLen);
+	#endif
+			strValue[nStrLen - 1] = TEXT('\0');
+
+			break;*/
+
+	default: result = std::string("Default Value");//std::cout << "No value for property" << std::endl;
+
+	}
+	VariantClear(&var);
+
+
+	return result;
+
+
+}
+
+template<typename Container>
+void D3DInfo<Container>::init(int dxVersion)
+{
+	HRESULT hr;
+
+	hr = CoInitialize(NULL);
+	throw_if_failed(hr, "Can not CoInitialize() COM");		
+
+	//m_bCleanupCOM = SUCCEEDED(hr);
+
+	hr = CoCreateInstance(CLSID_DxDiagProvider,
+		NULL,
+		CLSCTX_INPROC_SERVER,
+		IID_IDxDiagProvider,
+		(LPVOID*)&m_dxdiag_provider);
+
+	throw_if_failed(hr, "Can not CoCreateInstance CLSID_DxDiagProvider ");
+
+	/*if (m_dxdiag_provider == NULL)
+	{
+		hr = E_POINTER;
+		return hr;
+	}*/
+
+	// Fill out a DXDIAG_INIT_PARAMS struct and pass it to IDxDiagContainer::Initialize
+	// Passing in TRUE for bAllowWHQLChecks, allows dxdiag to check if drivers are 
+	// digital signed as logo'd by WHQL which may connect via internet to update 
+	// WHQL certificates.    
+	DXDIAG_INIT_PARAMS dxDiagInitParam;
+	ZeroMemory(&dxDiagInitParam, sizeof(DXDIAG_INIT_PARAMS));
+
+	dxDiagInitParam.dwSize = sizeof(DXDIAG_INIT_PARAMS);
+	dxDiagInitParam.dwDxDiagHeaderVersion = dxVersion;//DXDIAG_DX9_SDK_VERSION;
+	dxDiagInitParam.bAllowWHQLChecks = true;//bAllowWHQLChecks;
+	dxDiagInitParam.pReserved = NULL;
+
+	hr = m_dxdiag_provider->Initialize(&dxDiagInitParam);
+	throw_if_failed(hr, "Can not Initialize DXDIAG_INIT_PARAMS ");
+	//if (FAILED(hr))
+	//	return hr;
+
+	hr = m_dxdiag_provider->GetRootContainer(&m_dxdiag_root);
+	throw_if_failed(hr, "GetRootContainer from Provider Failed ");
+
+}
+
+template<typename Container>
+void D3DInfo<Container>::traverse(IDxDiagContainer * parentContainer, std::shared_ptr<Container> parent)
+{
+	IDxDiagContainer*	pChildContainer;
```

Comment:
Looks like some C-style code. In C++ we can declare variables where it is really neccessary